src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,194 @@
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { requestAsync } from '../../request';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = sbvrUtils;
+
+const VERSION = 1;
+
+interface LokiStream {
+	stream: {
+		device_id: string;
+		service_id: string;
+		is_system: string;
+		is_std_err: string;
+	};
+	values: Array<[string, string]>;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private webSockets: Map<string, WebSocket>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.webSockets = new Map();
+	}
+
+	public get available(): boolean {
+		// TODO: check /ready endpoint?
+		return true;
+	}
+
+	public async history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		const [response, body] = await requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/query_range`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'GET',
+			qs: {
+				query: this.getDeviceQuery(ctx),
+				limit: Number.isFinite(count) ? count : 1000,
+			},
+			json: true,
+		});
+		if (response.statusCode === 200) {
+			return this.fromLokiStreamsToDeviceLogs(body.data.result);
+		} else {
+			throw new BadRequestError(
+				`Query failed ${response.statusCode} "${response.body}"`,
+			);
+		}
+	}
+
+	public async publish(ctx: LogWriteContext, logs: DeviceLog[]): Promise<any> {
+		return requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/push`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'POST',
+			json: {
+				streams: logs.map((log) => this.toLokiStream(ctx, log)),
+			},
+		});
+	}
+
+	public subscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		if (!this.webSockets.has(key)) {
+			const tailUri =
+				`ws://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/tail` +
+				`?query=${encodeURIComponent(
+					this.getDeviceQuery(ctx),
+				)}&start=${Date.now()}000000`;
+			const ws = new WebSocket(tailUri, {
+				headers: { 'X-Scope-OrgID': ctx.belongs_to__application },
+			});
+			ws.on('message', (data: string) => {
+				try {
+					const messageObj = JSON.parse(data);
+					if (messageObj.streams) {
+						this.fromLokiStreamsToDeviceLogs(messageObj.streams).forEach(
+							(log: DeviceLog) => {
+								this.subscriptions.emit(key, log);
+							},
+						);
+					}
+				} catch (err) {
+					captureException(err, 'Unexpected message received from Loki tail');
+				}
+			});
+			ws.on('close', () => {
+				this.webSockets.delete(key);
+			});
+			this.webSockets.set(key, ws);
+		}
+		this.subscriptions.on(key, subscription);
+	}
+
+	public unsubscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		this.subscriptions.removeListener(key, subscription);
+		if (this.webSockets.has(key)) {
+			const ws = this.webSockets.get(key);
+			ws?.close();
+		}
```

Comment:
I think you can simplify this to
```suggestion
		const ws = this.webSockets.get(key);
		ws?.close();
```
and only look up the key once rather than once to check it exists and then again to actually retrieve it